Separate harm from sunlight to eyes and skin to albino characters

Night-Pryanik opened the pull request:
When wearing gear that protect eyes from sunlight and glare, including complete light blocking.
As per discussion in forum (http://smf.cataclysmdda.com/index.php?topic=12434.0)

Coolthulhu commented:
Don't pile those duplicate flag checks. They're hard to maintain. Wrap them up in functions, like `sunlight_bp_blocking` that checks a given body part for blocking. Or even better, make it return a bitset like the flag coverage functions.

Coolthulhu commented:
Albinism and especially trogglo shouldn't be just negated trivially with a pair of sunglasses. Not even welding goggles should be enough to negate full trogglo.

Albinism isn't just a matter of eyes being protected. For albino penalty not to hit at all, you'd need 100% coverage on all body parts affected by sun AND this coverage being only counted for "sunproof" items (not cotton or wool).

Sunlight hate hits all stats, not just intelligence and perception. It is also a mutant-exclusive trait, meaning it isn't something like just aversion to sun. So, as with albino, it needs to require a full body protection to negate.

Night-Pryanik commented:
We have separate `SUNBURN` mutation that affects body sensitivity to sunlight. So I thought that we have `ALBINO` and `TROGLO1-3` that irritates the eyes and `SUNBURN` that burns the skin. So body protection is needed only for `SUNBURN`.

From what I read wearing ordinary sunglasses IRL is enough to negate albino photosensitivity.

If eyes protection of welding glasses is enough to prevent eyes damage while welding (luminance of this process is much higher of that of sunlight if I'm not mistaken), then it is surely enough to prevent eyes from sunlight.

kevingranade commented:
Nothing says that the effects are exclusive, albinism and thevtroglodite
mutations are intended to be skin effects as well.

Night-Pryanik commented:
By the way, at the moment it is possible to negate penalties to albino, troglo and sunburn just by wearing motorcycle pants, even if other body parts are naked.

Coolthulhu commented:
Looks like someone added RAINPROOF to some items without first checking what does the flag do.

Only 3 items that don't cover the torso are rainproof:
* Motorcycle pants
* Light survivor cargo pants
* Motorcycle boots

Though even a full rainproof coat shouldn't fully negate those penalties. Umbrella could be enough for albino, but not for trog and sunburn.

Night-Pryanik commented:
I assume the logic is this (correct me if I'm wrong):

- albino is eyes
- sunburn is skin
- trog is albino + sunburn (= eyes + skin)

Making albino = eyes + skin is redundant IMO.
So protecting eyes with sunglasses (or similar) negates penalties for EYES ONLY for albino AND trog.

Coolthulhu commented:
Albinism affects the skin too. If we're removing one, I'd rather see all of them being skin only because it's much harder to protect the entire body and a negative trait worth 2+ points shouldn't be negated by just wearing sunglasses.

Night-Pryanik commented:
Albinos and photosensitive characters suffer from genetic problems (deficit of melanin and all that). I wonder what are the reasons for penalties for trogs? Some magic?

kevingranade commented:
Yes, some magic. Troglodite mutation is a reference to a fantasy race.

Night-Pryanik commented:
It seems that albino mutation still was meant for eyes only, because the check is 
`if( (has_trait("ALBINO") || has_effect( effect_datura ) )`
I don't know what does datura deal IRL, but I think that this is not related to skin, but to the eyes.

Coolthulhu commented:
The penalty blocks might have been merged at some point.

Albino starts with an umbrella rather than sunglasses. If it was eyes only, it would get the sunglasses, because umbrella is much bigger, has to be held and is rarer.

Also, if albino was negated by sunglasses, it wouldn't be worth even -1 points. It has -2 cost at the moment.

Night-Pryanik commented:
Well, we have more than enough traits that needs to be balanced. If we are talking about the eyes, there are near-sighted and far-sighted. They are both -2, and negated immediately after start by having glasses in inventory. One doesn't even need to do anything.
But finding sunglasses or hat with wide brims penalty of -2 turns out to be too much!
Personally I think being an albino with -2 points is too harsh for simply having minor pain and no other significant penalties.

Coolthulhu commented:
> Well, we have more than enough traits that needs to be balanced. If we are talking about the eyes, there are near-sighted and far-sighted. They are both -2, and negated immediately after start by having glasses in inventory. One doesn't even need to do anything.

Could be fixed by giving the glasses significant encumbrance. 10+

Also, it would be good if eyes had an encumbrance limit for stacking, to prevent nonsensical combos like glasses+goggles.

Possibly also removing the ability to repair glasses. Lenses aren't easy to repair.

But we're talking about albino. The argument that other trait is not worth its point value isn't enough here - you'd have to prove albino is worth the -2 points.

> But finding sunglasses or hat with wide brims penalty of -2 turns out to be too much!

Wide hat has effectively zero encumbrance because it is worn on the head, where encumbrance has no effect and helmet+boonie is possible.

If you want to make wide hats negate albino, make it not a starting trait. Having to wear a boonie hat is less of a penalty than lactose intolerance, which is considered barely worse than free point.

Rivet-the-Zombie commented:
>I don't know what does datura deal IRL, but I think that this is not related to skin, but to the eyes.

[Due to the potent combination of anticholinergic substances it contains, Datura intoxication typically produces effects similar to that of an anticholinergic delirium...  ...**and severe mydriasis (dilated pupils) with resultant painful photophobia that can last several days.**](https://en.wikipedia.org/wiki/Datura#Effects_of_ingestion)

In other words, the resulting photosensitivity is specifically confined to the eyes.

Night-Pryanik commented:
I think we need to come to some sort of consensus here by answering these questions.
1) Will albino affect eyes only, or eyes and skin?
2) Is -2 worth of minor pain?

Coolthulhu commented:
> Is -2 worth of minor pain?

Nope. Except maybe if it was nearly constant, very hard to prevent.

kevingranade commented:
1) eyes and skin
2) no

Night-Pryanik commented:
Made some work:
- Separated ALBINO into eyes and skin, either body part must have its own protection. Only eyegear with glare protection or light blocking will work, not other gear with SUN_GLASSES flag, like boonie hat. To fully neglect ALBINO penalties, one should wield umbrella AND wear sunglasses or blindfold.
- Removed check for gear with RAINPROOF flag for ALBINO and SUNBURN mutations.
- Reverted checks to original for TROGLO1-3, as they need more complex checks like item material and percent of body coverage.

kevingranade commented on src/player.cpp:
```diff
@@ -7919,24 +7919,26 @@ void player::suffer()
+    if( ( has_trait( "ALBINO" ) || has_effect( effect_datura ) ) &&
+        g->is_in_sunlight( pos() ) && one_in(10) ) {
+        // Umbrellas can keep the sun off the skin and sunglasses - off the eyes.
+        if( !weapon.has_flag( "RAIN_PROTECT" ) ) {
```
The nesting is wrong now, having the protective items will keep the messages from being displayed, but you'd still be woken up and suffer pain.

kevingranade commented on data/json/mutations.json:
```diff
@@ -870,8 +870,8 @@
-        "points" : -2,
-        "description" : "You lack skin pigmentation due to a genetic problem.  You sunburn extremely easily, and typically use an umbrella when going out in the sun.",
+        "points" : -1,
+        "description" : "You lack skin pigmentation due to a genetic problem.  You sunburn extremely easily, and typically use an umbrella and a sunglagges when going out in the sun.",
```
sunglasses

kevingranade commented on src/player.cpp:
```diff
@@ -7919,24 +7919,26 @@ void player::suffer()
+    if( ( has_trait( "ALBINO" ) || has_effect( effect_datura ) ) &&
+        g->is_in_sunlight( pos() ) && one_in(10) ) {
+        // Umbrellas can keep the sun off the skin and sunglasses - off the eyes.
+        if( !weapon.has_flag( "RAIN_PROTECT" ) ) {
```
Now there's no penalty for skin exposure, just a message.
